### PR TITLE
MM with persistant msg_ids; DM with more robust mdata refresh

### PR DESCRIPTION
**Pull request by maqi**

### CODE REVIEW

File: src/mock_routing.rs [hunk 1]
```diff
@@ -226,14 +226,13 @@ impl Node {
                     dst: Authority<XorName>,
                     request: Request)
                     -> Result<(), InterfaceError> {
-        let prev = self.sent_requests
+        let _ = self.sent_requests
```
Comment (madadam) [inline src/mock_routing.rs]: What's is the reason for this change?

File: src/vault.rs [hunk 2]
```diff
@@ -760,3 +779,9 @@ enum EventResult {
     // `RestartRequired` event received.
     Restart,
 }
+
+#[derive(Deserialize, Serialize, PartialEq, Eq, Debug, Clone)]
+pub enum Refresh {
```
Comment (madadam) [inline src/vault.rs:781]: Both `maid_manager` and `data_manager` modules now define their own refresh types, so I'd prefer this to be changed to:
```rust
pub enum Refresh {
    MaidManager(maid_manager::Refresh),
    DataManager(data_manager::Refresh),
}
```
so we could avoid the double serialization.

File: src/mock_routing.rs [hunk 3]
```diff
@@ -242,14 +241,13 @@ impl Node {
                      dst: Authority<XorName>,
                      response: Response)
                      -> Result<(), InterfaceError> {
-        let prev = self.sent_responses
+        let _ = self.sent_responses
```
Comment (madadam) [inline src/mock_routing.rs]: What's the reason for this change?

**Review by madadam — CHANGES_REQUESTED** [review]

File: src/vault.rs (hunk 2 above)
Comment (maqi) [reply src/vault.rs:781]: changed

File: src/personas/maid_manager/mod.rs [hunk 4]
```diff
@@ -16,64 +16,80 @@
...
     pub fn new() -> MaidManager {
         MaidManager {
             accounts: HashMap::default(),
+            data_ops_msg_id_accumulator:
+                MessageIdAccumulator::new(QUORUM, Duration::from_secs(ACCUMULATOR_TIMEOUT_SECS)),
             request_cache: HashMap::default(),
         }
     }
 
+    pub fn handle_serialised_refresh(&mut self,
```
Comment (madadam) [inline src/personas/maid_manager/mod.rs:59]: Use the same naming convention in both `MaidManager` and `DataManager`. 
- Either `handle_serialised_refresh` and `handle_refresh`
- Or `handle_refresh` and `handle_parsed_refresh`. 

For consistency.

File: src/personas/maid_manager/mod.rs [hunk 5]
```diff
@@ -703,23 +858,69 @@ impl MaidManager {
...
             .get(client_name)
-            .map(|account| account.info.mutations_done)
+            .map(|account| account.data_ops_msg_ids.len() as u64)
     }
 }
 
 #[derive(Deserialize, Serialize, PartialEq, Eq, Debug, Clone)]
-enum Refresh {
-    Update(XorName, Account),
+pub enum Refresh {
```
Comment (madadam) [inline src/personas/maid_manager/mod.rs:866]: The names of the variants are now confusing and do not reflect well their purpose. My suggestion:
- rename `Update` to `UpdateDataOps`
- rename `UpdateData` to `InsertDataOp`

Also for consistency, rename the methods:

- `handle_refresh_update` to `handle_refresh_update_data_ops`
- `handle_refresh_update_data` to `handle_refresh_insert_data_op`

Feel free to suggest different names, but the important thing is to make them descriptive and consistent.

**Review by madadam — CHANGES_REQUESTED** [review]

File: src/personas/maid_manager/mod.rs (hunk 4 above)
Comment (maqi) [reply src/personas/maid_manager/mod.rs:59]: changed

File: src/personas/maid_manager/mod.rs (hunk 5 above)
Comment (maqi) [reply src/personas/maid_manager/mod.rs:866]: changed

File: src/personas/data_manager/cache.rs [hunk 6]
```diff
@@ -56,17 +65,90 @@ impl Cache {
...
+    /// Mark the request to retrieve the given chunk as started.
+    pub fn start_needed_mutable_chunk_request(&mut self,
+                                              data_id: MutableDataId,
+                                              msg_id: MessageId) {
+        let _ = self.needed_mutable_chunks.remove(&data_id);
+        self.needed_mutable_chunk_request = Some(ChunkRequest::new(data_id, msg_id));
+    }
+
+    /// Register successful response to a needed mutable chunk request. If we receive
+    /// responses from at least QUORUM nodes, the request is considered to be successfuly
```
Comment (afck) [inline src/personas/data_manager/cache.rs]: `successfully` and ``` `QUORUM` ```

File: src/personas/data_manager/cache.rs [hunk 7]
```diff
@@ -77,24 +159,15 @@ impl Cache {
...
-                    _ => {
-                        if self.fragment_index
-                               .get(fragment)
-                               .map_or(false, |state| state.is_requested()) {
-                            continue;
-                        }
-                    }
+                if self.fragment_index
+                       .get(fragment)
+                       .map_or(false, |state| state.is_requested()) {
```
Comment (afck) [inline src/personas/data_manager/cache.rs]: Is this closure needed? Could probably just be `State::is_requested`?

File: src/personas/data_manager/cache.rs [hunk 8]
```diff
@@ -56,17 +65,90 @@ impl Cache {
...
+
+        if done {
+            self.needed_mutable_chunk_request = None;
+        }
+    }
+
     /// Returns data fragments we need and want to request from other nodes.
     pub fn needed_fragments(&mut self) -> Vec<(FragmentInfo, Vec<XorName>)> {
-        self.stop_expired_requests();
+        self.stop_expired_fragment_requests();
```
Comment (afck) [inline src/personas/data_manager/cache.rs:155]: Should this method really iterate through _all_ holders and return a vector of all needed fragments? This could be a long iteration and we're probably only interested in the next bunch of requests we need to send?

File: src/personas/data_manager/cache.rs [hunk 9]
```diff
@@ -681,6 +756,38 @@ impl FragmentState {
...
+            timestamp: Instant::now(),
+        }
+    }
+
+    fn is_expired(&self) -> bool {
+        self.timestamp.elapsed().as_secs() > MUTABLE_CHUNK_REQUEST_TIMEOUT_SECS
+    }
+
+    fn can_accumulate(&self) -> bool {
+        GROUP_SIZE - 1 - self.failures.len() >= QUORUM
```
Comment (afck) [inline src/personas/data_manager/cache.rs]: Maybe add a comment that that `1` represents our own node.

File: src/personas/data_manager/cache.rs [hunk 10]
```diff
@@ -829,15 +936,15 @@ mod tests {
...
+        let fragment = FragmentInfo::ImmutableData(rng.gen());
         let holder0 = rng.gen();
         let holder1 = rng.gen();
 
         assert!(cache.fragment_holders.is_empty());
         assert!(cache.fragment_index.is_empty());
 
         // Start multiple requests, then stop them. Assert that the needed fragment
-        // data strucutres ramain empty afterwards.
+        // data strucutres remains empty afterwards.
```
Comment (afck) [inline src/personas/data_manager/cache.rs]: `structure`

File: src/personas/data_manager/mod.rs [hunk 11]
```diff
@@ -1195,10 +1154,80 @@ impl DataManager {
...
+    fn handle_fragment_refresh(&mut self,
+                               src: XorName,
+                               fragment: FragmentInfo)
+                               -> Result<(), InternalError> {
+        if self.cache
+               .register_needed_fragment_with_another_holder(fragment.clone(), src) {
+            return Ok(());
+        }
+
+        let holders = match self.fragment_refresh_accumulator
```
Comment (afck) [inline src/personas/data_manager/mod.rs:1177]: Are the holders relevant in the case of mutable data?

File: src/personas/data_manager/tests.rs [hunk 12]
```diff
@@ -730,33 +726,35 @@ fn setup_mdata_refresh<R: Rng>(data: &MutableData,
...
+    });
+
     let (name, tag) =
         assert_match!(message.request, Request::GetMData { name, tag, .. } => (name, tag));
-    (MutableDataId(name, tag), message.src)
+    (msg_id, MutableDataId(name, tag))
 }
 
-fn take_request<F>(node: &mut RoutingNode, mut f: F) -> (RequestWrapper)
+fn take_request<F>(node: &mut RoutingNode, mut f: F) -> (MessageId, RequestWrapper)
```
Comment (afck) [inline src/personas/data_manager/tests.rs]: It would be helpful to have a comment that explains that `f` is a filter function here.

File: src/personas/data_manager/cache.rs (hunk 8 above)
Comment (madadam) [reply src/personas/data_manager/cache.rs:155]: So what's your suggestion? Should we limit the maximum number of items returned from this function or something like that?

Comment (afck) [reply src/personas/data_manager/cache.rs:155]: Yes, wouldn't it be sufficient to return the first needed fragment for each data holder?
Not sure, though. If it's a big change that could be left as an optimisation for later.

File: src/personas/data_manager/mod.rs (hunk 11 above)
Comment (madadam) [reply src/personas/data_manager/mod.rs:1177]: Yes, because this function is used for: 
- refreshing immutable data after node joined/lost.
- refreshing immutable or mutable data after mutation. In case of mutable data, only the modified parts of it are refreshed.

File: src/personas/data_manager/mod.rs [hunk 13]
```diff
@@ -330,33 +286,37 @@ impl DataManager {
...
         self.update_request_stats(&src);
 
+        /// Send the response as a node, to allow custom accumulation.
+        let resp_src = Authority::ManagedNode(*routing_node.id()?.name());
+        let resp_dst = src;
+
         let res = self.fetch_mdata(name, tag);
         trace!("As {:?} sending GetMData response {:?} of {:?} to {:?}",
-               dst,
+               resp_dst,
```
Comment (maqi) [inline src/personas/data_manager/mod.rs]: shall here be resp_src?

File: src/personas/data_manager/mod.rs [hunk 14]
```diff
@@ -330,33 +286,37 @@ impl DataManager {
...
                res,
                name,
-               src);
+               resp_src);
```
Comment (maqi) [inline src/personas/data_manager/mod.rs]: shall here be resp_dst?

File: src/personas/data_manager/cache.rs (hunk 8 above)
Comment (madadam) [reply src/personas/data_manager/cache.rs:155]: Yeah, not sure too. I'd leave this for later. I think we need an optimization pass anyway, but only after careful profiling.

File: src/personas/data_manager/mod.rs [hunk 15]
```diff
@@ -1195,10 +1154,80 @@ impl DataManager {
...
+                name,
+                tag,
+                ref key,
+                version,
+                ..
+            } => {
+                if let Ok(data) = self.chunk_store.get(&MutableDataId(name, tag)) {
+                    match data.get(key) {
+                        None => true,
+                        Some(value) if value.entry_version < version => true,
```
Comment (maqi) [inline src/personas/data_manager/mod.rs]: why not Some(value)  => value.entry_version < version,
then don't need the match arm of Some(_) => false ?
